### PR TITLE
Ceph object store for external cluster

### PR DESCRIPTION
**Pull request by aruniiird**
CephObjectStore creation in external cluster mode according to the latest rook-ceph changes

### CODE REVIEW

**Comment by aruniiird** [comment]
@leseb , can you please take a look.

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 1]
```diff
@@ -374,6 +384,50 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 	return nil
 }
 
+// newExternalCephObjectStoreInstances returns a set of CephObjectStores
+// needed for external cluster mode
+func (r *ReconcileStorageCluster) newExternalCephObjectStoreInstances(
+	initData *ocsv1.StorageCluster, rgwEndpoint string, reqLogger logr.Logger) ([]*cephv1.CephObjectStore, error) {
+	rgwEndpoint = strings.TrimSpace(rgwEndpoint)
+	var gateWay cephv1.GatewaySpec
+	if rgwEndpoint != "" {
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:393]: I don't think it can be empty, the UI will not accept that, or perhaps are you trying to catch a manual deployment issue?

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 2]
```diff
@@ -205,9 +208,14 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
...
+				if err != nil {
+					return err
+				}
 				// Set the external rgw endpoint variable for later use on the Noobaa CR (as a label)
 				// Replace the colon with an underscore, otherwise the label will be invalid
-				externalRgwEndpointReplaceColon := strings.Replace(d.Data[externalCephRgwEndpointKey], ":", "_", -1)
+				externalRgwEndpointReplaceColon := strings.Replace(rgwEndpoint, ":", "_", -1)
 				externalRgwEndpoint = externalRgwEndpointReplaceColon
 
 				// Setting the Endpoint for OBC StorageClass
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:221]: We don't need to set the endpoint anymore on the SC for OBC. We just need to set the `objectStoreName` back to it, just like we do for the converged case.

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 1 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go:393]: @leseb , there was discussion regarding making the `rgw-endpoint` optional and this function should serve both the needs.

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 2 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go]: Made the changes, @leseb , please take a look...

**Comment by aruniiird** [comment]
Rebased this PR on top of @umangachapagain 's [Update rook to 1.4](https://github.com/openshift/ocs-operator/pull/657) PR.
Removed all the `TODO` s to working code. @leseb , please take a look...

**Comment by aruniiird** [comment]
This PR depends on https://github.com/openshift/ocs-operator/pull/657

**Comment by aruniiird** [comment]
@anmolsachan , this is the PR for the new CephObjectStore changes. Please take a look...

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 3]
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
...
+	var gateWay cephv1.GatewaySpec
+	hostIP, portStr, err := net.SplitHostPort(rgwEndpoint)
+	if err != nil {
+		reqLogger.Error(err,
+			fmt.Sprintf("invalid rgw endpoint provided: %s", rgwEndpoint))
+		return nil, err
+	}
+	if hostIP == "" {
+		err := fmt.Errorf("An empty rgw host 'IP' address found")
+		reqLogger.Error(err, err.Error())
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:407]: Why not just `reqLogger.Error(err)`?

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 4]
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 	return nil
 }
 
+// newExternalCephObjectStoreInstances returns a set of CephObjectStores
+// needed for external cluster mode
+func (r *ReconcileStorageCluster) newExternalCephObjectStoreInstances(
+	initData *ocsv1.StorageCluster, rgwEndpoint string, reqLogger logr.Logger) ([]*cephv1.CephObjectStore, error) {
+	// check whether the provided rgw endpoint is empty
+	if rgwEndpoint = strings.TrimSpace(rgwEndpoint); rgwEndpoint == "" {
+		reqLogger.Info("WARNING: Empty RGW Endpoint specified, external CephObjectStore won't be created")
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:394]: Is it a warning or an error? Either way, use the appropriate `reqLogger`.

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 5]
```diff
@@ -374,6 +385,59 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
...
+		return nil, nil
+	}
+
+	var gateWay cephv1.GatewaySpec
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:398]: What about a method that returns a `cephv1.GatewaySpec` instead of this long method? It will be easier to unit test too.

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 6]
```diff
@@ -145,6 +147,7 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
 		Kind:       instance.Kind,
 		Name:       instance.Name,
 	}
+	var extCephObjectStores []*cephv1.CephObjectStore
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:150]: Why an array?

File: pkg/controller/storagecluster/generate.go [hunk 7]
```diff
@@ -27,7 +27,12 @@ func generateNameForCephBlockPool(initData *ocsv1.StorageCluster) string {
 }
 
 func generateNameForCephObjectStore(initData *ocsv1.StorageCluster) string {
```
Comment (leseb) [inline pkg/controller/storagecluster/generate.go:29]: Do you have unit tests for this?

File: pkg/controller/storagecluster/initialization_reconciler.go [hunk 8]
```diff
@@ -205,24 +208,36 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
...
+				if err != nil {
+					return err
+				}
+				// rgw-endpoint is no longer needed in the 'd.Data' dictionary,
+				// and can be deleted
+				delete(d.Data, externalCephRgwEndpointKey)
 				// Set the external rgw endpoint variable for later use on the Noobaa CR (as a label)
 				// Replace the colon with an underscore, otherwise the label will be invalid
-				externalRgwEndpointReplaceColon := strings.Replace(d.Data[externalCephRgwEndpointKey], ":", "_", -1)
+				externalRgwEndpointReplaceColon := strings.Replace(rgwEndpoint, ":", "_", -1)
```
Comment (leseb) [inline pkg/controller/storagecluster/initialization_reconciler.go:221]: I need to check with @nb-ohad, for changes in Noobaa but this might need to go away as well.

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 4 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go:394]: Yes it is a `Warning` only and, in a normal scenario, this function won't be even called (as won't be propagating the `ceph-rgw` in the json output).
`reqLogger` (an instance of `github.com/go-logr/logr` interface) don't have a `.Warn` (and have a [philosophy](https://dave.cheney.net/2015/11/05/lets-talk-about-logging) on why no warnings). So `Info` is the best (and non-complicated) bet to go.

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 6 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go:150]: This is to pass it to `createCephObjectStoreInstances` method (which takes an array of `*cephv1.CephObjectStore` instances as first argument).

File: pkg/controller/storagecluster/generate.go (hunk 7 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/generate.go:29]: Not yet, will add one...

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 3 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go:407]: `reqLogger.Error` requires a second string argument, which provides a message. Will change to a different message (rather than the same error message).

**Comment by leseb** [comment]
@aruniiird please rebase

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 9]
```diff
@@ -26,16 +26,14 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 		return nil
 	}
 
-	cephObjectStores, err := r.newCephObjectStoreInstances(instance)
+	var cephObjectStores []*cephv1.CephObjectStore
+	if cephObjectStores, err = r.newCephObjectStoreInstances(instance); err == nil {
+		err = r.createCephObjectStores(cephObjectStores, instance, reqLogger)
+	}
 	if err != nil {
+		reqLogger.Error(err, "could not create CephObjectStores")
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephobjectstores.go]: These changes don't seem related to the commit message. This can probably be integrated into the next commit.

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 10]
```diff
@@ -28,7 +28,7 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
 
 	var cephObjectStores []*cephv1.CephObjectStore
 	if cephObjectStores, err = r.newCephObjectStoreInstances(instance); err == nil {
-		err = r.createCephObjectStores(cephObjectStores, instance, reqLogger)
+		err = r.createCephObjectStores(cephObjectStores, reqLogger)
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephobjectstores.go:31]: Why is this being removed here? Was it never used?

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 11]
```diff
@@ -38,28 +38,35 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
...
 		switch {
 		case err == nil:
 			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				err := fmt.Errorf("failed to restore initialization object %s because it is marked for deletion",
+					existing.Name)
+				reqLogger.Error(err,
+					fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
+				return err
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephobjectstores.go:52]: Why are there so many error message changes? These should be in a separate commit. Also, please don't wrap the fmt/logging function lines.

File: pkg/controller/storagecluster/external_resources.go [hunk 12]
```diff
@@ -141,9 +205,17 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterResources(scs []*s
...
 				sc = scs[1]
 			} else if d.Name == cephRgwStorageClassName {
+				rgwEndpoint := d.Data[externalCephRgwEndpointKey]
+				extCephObjectStores, err = r.newExternalCephObjectStoreInstances(instance, rgwEndpoint, reqLogger)
+				if err != nil {
+					return err
+				}
+				// rgw-endpoint is no longer needed in the 'd.Data' dictionary,
+				// and can be deleted
+				delete(d.Data, externalCephRgwEndpointKey)
```
Comment (jarrpa) [inline pkg/controller/storagecluster/external_resources.go:215]: What's the point of doing this? What happens if you don't delete it?

File: pkg/controller/storagecluster/external_resources.go [hunk 13]
```diff
@@ -77,15 +76,86 @@ func (r *ReconcileStorageCluster) ensureExternalStorageClusterResources(instance
...
+			fmt.Sprintf("invalid rgw endpoint provided: %s", rgwEndpoint))
+		return nil, err
+	}
+	if hostIP == "" {
+		err := fmt.Errorf("An empty rgw host 'IP' address found")
+		reqLogger.Error(err, "Host IP should not be empty in rgw endpoint")
+		return nil, err
+	}
+	gateWay.ExternalRgwEndpoints = []corev1.EndpointAddress{{IP: hostIP}}
+	if portInt64, err := strconv.ParseInt(portStr, 10, 32); err != nil {
```
Comment (anmolsachan) [inline pkg/controller/storagecluster/external_resources.go:105]:
```suggestion
	portInt64, err := strconv.ParseInt(portStr, 10, 32)
	if err != nil {
		reqLogger.Error(err,
			fmt.Sprintf("invalid rgw 'port' provided: %s", portStr))
		return nil, err
	}
	gateWay.Port = int32(portInt64)
```

Comment (anmolsachan) [reply pkg/controller/storagecluster/external_resources.go:105]: Can you please refractor it a little? It's giving lint error

File: pkg/controller/storagecluster/initialization_reconciler.go (hunk 5 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/initialization_reconciler.go]: Done...

File: pkg/controller/storagecluster/generate.go (hunk 7 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/generate.go:29]: Added unit tests...

File: pkg/controller/storagecluster/cephobjectstores.go (hunk 9 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go]: Moved to a separate commit...

File: pkg/controller/storagecluster/cephobjectstores.go (hunk 10 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go:31]: It was never used in the `createCephObjectStores()` function.

File: pkg/controller/storagecluster/cephobjectstores.go (hunk 11 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go:52]: Corrected the error message and moved to the refactoring commit...

File: pkg/controller/storagecluster/external_resources.go (hunk 12 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/external_resources.go:240]: RGW endpoint details we already stored in a `CephObjectStore`. If you don't `delete()` , the rgwEndpoint key / value will be again added to the StorageClass' `Parameters` map, which is not needed.

In the previous design, we used to store the rgw-endpoint details in the StorageClass itself, but now rgw-endpoint details are stored in a `CephObjectStore` object and in the SC, as one of the parameter, we just point to this cephobjectstore's name.

File: pkg/controller/storagecluster/external_resources.go (hunk 13 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/external_resources.go]: Corrected...

**Comment by aruniiird** [comment]
Added test cases... Ready for final reviews ... @jarrpa , @leseb please see

File: pkg/controller/storagecluster/external_resources.go [hunk 14]
```diff
@@ -205,16 +282,20 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterConfigMap(cm *core
 			err = r.client.Create(context.TODO(), cm)
 			if err != nil {
 				reqLogger.Error(err, "creation of configmap failed")
-				return err
 			}
 		} else {
 			reqLogger.Error(err, "unable the get the configmap")
-			return err
 		}
+		return err
```
Comment (jarrpa) [inline pkg/controller/storagecluster/external_resources.go:289]: Don't do this, leave the return statements as-is.

File: pkg/controller/storagecluster/external_resources.go [hunk 15]
```diff
@@ -205,16 +282,20 @@ func (r *ReconcileStorageCluster) createExternalStorageClusterConfigMap(cm *core
...
 	}
 	return nil
 }
 
+/*
+go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20200302205851-738671d3881b
+./pkg/controller/storagecluster/external_resources.go:109:9: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
+*/
+
```
Comment (jarrpa) [inline pkg/controller/storagecluster/external_resources.go:298]: Is this supposed to be here?

File: pkg/controller/storagecluster/external_resources.go (hunk 12 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/external_resources.go:240]: ...so then why is this still being passed in the RGW StorageClass data? There should be a separate place for it so that this hack isn't necessary.

File: pkg/controller/storagecluster/storageclasses.go [hunk 16]
```diff
@@ -84,14 +84,9 @@ func (r *ReconcileStorageCluster) newOBCStorageClass(initData *ocsv1.StorageClus
 		Parameters: map[string]string{
 			"objectStoreNamespace": initData.Namespace,
 			"region":               "us-east-1",
+			"objectStoreName":      generateNameForCephObjectStore(initData),
```
Comment (jarrpa) [inline pkg/controller/storagecluster/storageclasses.go:87]: So now we're generating this for both external and converged modes? Did something that drastic really change in the OBC StorageClass?

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 17]
```diff
@@ -55,12 +55,16 @@ func (r *ReconcileStorageCluster) createCephObjectStores(cephObjectStores []*cep
...
+				reqLogger.Error(err,
+					fmt.Sprintf("Failed to update CephObjectStore Object: %s", cephObjectStore.Name))
 				return err
 			}
 		case errors.IsNotFound(err):
 			reqLogger.Info(fmt.Sprintf("Creating CephObjectStore %s", cephObjectStore.Name))
 			err = r.client.Create(context.TODO(), cephObjectStore)
 			if err != nil {
+				reqLogger.Error(err,
+					fmt.Sprintf("Creation failed for CephObjectStore object: %s", cephObjectStore.Name))
```
Comment (jarrpa) [inline pkg/controller/storagecluster/cephobjectstores.go:67]: Move all changes to this file in this second commit to the first commit. These are part of the refactoring and do not impact the addition of the external CephObjectStore.

**Comment by jarrpa** [comment]
Holding until #657 merges.

/hold

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 18]
```diff
@@ -26,16 +26,14 @@ func (r *ReconcileStorageCluster) ensureCephObjectStores(instance *ocsv1.Storage
...
 	if err != nil {
+		reqLogger.Error(err, "could not create CephObjectStores")
 		return err
 	}
-
-	err = r.createCephObjectStores(cephObjectStores, instance, reqLogger)
-	if err != nil {
-                reqLogger.Error(err,"could not create CephObjectStores")
-                return err
-        }
```
[the remark this answers is not in the document]
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go]: Done...

File: pkg/controller/storagecluster/cephobjectstores.go [hunk 19]
```diff
@@ -47,8 +45,9 @@ func (r *ReconcileStorageCluster) createCephObjectStores(cephObjectStores []*cep
 		switch {
 		case err == nil:
 			if existing.DeletionTimestamp != nil {
-				reqLogger.Info(fmt.Sprintf("Unable to restore init object because %s is marked for deletion", existing.Name))
-				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+				err := fmt.Errorf("failed to restore cephobjectstore object %s because it is marked for deletion", existing.Name)
+				reqLogger.Error(err, "cephobjectstore restore failed")
```
[the remark this answers is not in the document]
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go]: Done...

File: pkg/controller/storagecluster/external_resources.go (hunk 14 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/external_resources.go]: Yup, changed it ...

File: pkg/controller/storagecluster/external_resources.go (hunk 15 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/external_resources.go]: Removed it...

File: pkg/controller/storagecluster/storageclasses.go (hunk 16 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/storageclasses.go:87]: A minor change in the OBC StorageClass' `Parameter map[string]string` now takes `objectStoreName`, instead of direct `endpoint`.

File: pkg/controller/storagecluster/cephobjectstores.go (hunk 17 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/cephobjectstores.go]: Done...

File: pkg/controller/storagecluster/external_resources.go (hunk 12 above)
Comment (aruniiird) [reply pkg/controller/storagecluster/external_resources.go:240]: What I understood was,
We didn't make any change in the rook's JSON external input creation script (and kept it backward compatible). Thus changes were done only on OCS-Operator side with the same receiving input file.

**Comment by aruniiird** [comment]
@jarrpa , addressed all the review comments... Please take a look...

File: pkg/controller/storagecluster/storageclasses.go (hunk 16 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/storageclasses.go:87]: I'd hardly call this minor. :) Also, you didn't answer my question: this is for both external and converged modes? How does it get used in converged mode, and how did it work before without this?

Comment (aruniiird) [reply pkg/controller/storagecluster/storageclasses.go:87]: Oops! missed the question mark (read the first line as a statement not as question)...
Yes, we are generating the CephObjectStore for both external and converged mode... But the creation of `CephObjectStore` is different in each mode,
a. Converged mode - we use ensure function to create CephObjectStore
b. External mode - we create it when `ceph-rgw` storageclass is available with the endpoint data in it.

@leseb ^^^

File: pkg/controller/storagecluster/external_resources.go (hunk 12 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/external_resources.go:242]: Thank you!!
